enterprise/server/scheduling/scheduler_server/scheduler_server.go

```diff
@@ -419,14 +420,26 @@ func (h *executorHandle) EnqueueTaskReservation(ctx context.Context, req *scpb.E
 	req = req.CloneVT()
 	tracing.InjectProtoTraceMetadata(ctx, req.GetTraceMetadata(), func(m *tpb.Metadata) { req.TraceMetadata = m })
 
-	// Just before enqueueing, resize the task to match the measured or
-	// predicted task size, up to the executor's limits. This late-resizing
-	// approach ensures that we always determine schedulability using the
-	// "default" estimate, which is stable. This way, tasks can't suddenly
-	// become unschedulable due to a change in measurements or an adjustment to
-	// the prediction model.
+	if req.GetSchedulingMetadata() == nil {
+		return nil, status.InvalidArgumentError("request is missing scheduling metadata")
+	}
+
+	// At this point, we haven't yet used the measured size or predicted size at
+	// all when deciding how to schedule this task. This is intentional - we
+	// don't want a task to suddenly become unschedulable on some nodes in a
+	// heterogenous pool, just because it executed one time on one of the larger
+	// nodes, and it greedily used up lots of resources on that node.
+	//
+	// So, once we've decided to schedule a task on a node, only then do we
+	// apply the measured and predicted sizes - but we limit to the node's
+	// capacity, to ensure that it is still schedulable within the node.
```

Comment:
I understand this comment better now, thanks for clarifying it a bit.

I still find the name `adjustTaskSize` very confusing -- because what I believe is happening is:
 - we compute several different task sizes using different methods
 - when selecting executors to run a task, we use the default estimate, not the measured or predicted size (right?)
 - when running a task on a node, we use the best size (measured or predicted or default or something)

I mentally think of the full code as doing:
executors := filterExecutors(allExecutors, task.GetDefaultSize())
...
mostAccurateTaskSize := selectMostAccurateTaskSize(task)
req.SchedulingMetadata.CgroupSettings = GetCgroupSettings(mostAccurateTaskSize)
...

So it's a bit confusing that "adjustTaskSize" is called adjust (what is being adjusted??) and that it mutates req, rather than just selecting the best size or something.

Reply:
I agree that the "adjust" terminology is confusing - changed it to a function that returns the most accurate size like in your example. We still need to set `req.TaskSize` and `req.SchedulingMetadata.TaskSize` though, because those fields determine the actual size on the executor. (If that part is still confusing, that's part of the reason why I thought we should deprecate `req.SchedulingMetadata.TaskSize` - though I think it could still be clearer)